File: studio/AuraStudio/Tests/AuraStudioTests/LibraryGroupingTests.swift
import XCTest
@testable import AuraStudio

/// ST-031 (PLAN-studio-ux.md §2.3): agrupación por álbum y por artista
/// -- casos obligatorios del plan: vacíos, homónimos, normalización,
/// "Sin álbum"/"Artista desconocido" al final.
final class LibraryGroupingTests: XCTestCase {
    private func song(_ title: String, artist: String? = nil, albumArtist: String? = nil,
                      album: String? = nil, track: Int? = nil, disc: Int? = nil,
                      year: String? = nil, cover: Data? = nil, kind: String = "mp3") -> AuraStudio.LibraryItem {
        var item = AuraStudio.LibraryItem(sourceURL: URL(fileURLWithPath: "/tmp/\(UUID().uuidString).\(kind)"))
        item.metadata = TrackMetadata(title: title, artist: artist, album: album, albumArtist: albumArtist,
                                      year: year, trackNumber: track, coverArtData: cover, discNumber: disc)
        return item
    }

    func testEmptyLibraryYieldsNoGroups() {
        XCTAssertTrue(LibraryGrouping.albums(from: []).isEmpty)
        XCTAssertTrue(LibraryGrouping.artists(from: []).isEmpty)
    }

    func testAlbumsIgnoreNonMusicItems() {
        let photo = AuraStudio.LibraryItem(sourceURL: URL(fileURLWithPath: "/tmp/x.jpg"))
        XCTAssertTrue(LibraryGrouping.albums(from: [photo]).isEmpty)
    }

    func testSameAlbumDifferentSpellingCollapsesIntoOneGroup() {
        let a = song("Uno", artist: "Café Tacvba", album: "Re")
        let b = song("Dos", artist: "Cafe Tacvba", album: " re ")
        let groups = LibraryGrouping.albums(from: [a, b])
        XCTAssertEqual(groups.count, 1)
        XCTAssertEqual(groups[0].title, "Re", "se muestra la grafía de la primera pista")
        XCTAssertEqual(groups[0].trackCount, 2)
    }

    func testHomonymousAlbumsOfDifferentArtistsStaySeparate() {
        let a = song("x", artist: "A", album: "Greatest Hits")
        let b = song("y", artist: "B", album: "Greatest Hits")
        XCTAssertEqual(LibraryGrouping.albums(from: [a, b]).count, 2)
    }

    func testAlbumArtistTakesPrecedenceOverTrackArtist() {
        let a = song("Feat 1", artist: "Gorillaz feat. Daley", albumArtist: "Gorillaz", album: "Cracker Island")
        let b = song("Feat 2", artist: "Gorillaz", albumArtist: "Gorillaz", album: "Cracker Island")
        let albums = LibraryGrouping.albums(from: [a, b])
        XCTAssertEqual(albums.count, 1)
        XCTAssertEqual(albums[0].artist, "Gorillaz")
        let artists = LibraryGrouping.artists(from: [a, b])
        XCTAssertEqual(artists.map(\.name), ["Gorillaz"])
    }

    func testUnknownAlbumGoesLastAndUsesLabel() {
        let known = song("k", artist: "Zeta", album: "Zzz")
        let noAlbum = song("n", artist: "Alfa")
        let groups = LibraryGrouping.albums(from: [noAlbum, known])
        XCTAssertEqual(groups.map(\.title), ["Zzz", LibraryGrouping.unknownAlbumTitle])
        XCTAssertTrue(groups[1].isUnknown)
        XCTAssertEqual(groups[1].artist, "Alfa")
    }

    func testAllWithoutAlbumIsOneGroupPerArtist() {
        let a = song("1", artist: "Alfa"), b = song("2", artist: "Alfa"), c = song("3", artist: "Beta")
        let groups = LibraryGrouping.albums(from: [a, b, c])
        XCTAssertEqual(groups.count, 2)
        XCTAssertTrue(groups.allSatisfy { $0.isUnknown })
        XCTAssertEqual(groups.map(\.artist), ["Alfa", "Beta"])
    }

    func testUnknownArtistGoesLast() {
        let a = song("1", album: "Solo")
        let b = song("2", artist: "Zzz", album: "Top")
        let artists = LibraryGrouping.artists(from: [a, b])
        XCTAssertEqual(artists.map(\.name), ["Zzz", LibraryGrouping.unknownArtistName])
        XCTAssertTrue(artists[1].isUnknown)
    }

    func testTracksSortByDiscThenTrackThenTitle() {
        let t3 = song("c", artist: "A", album: "X", track: 3, disc: 1)
        let d2 = song("a", artist: "A", album: "X", track: 1, disc: 2)
        let t1 = song("b", artist: "A", album: "X", track: 1, disc: 1)
        let noNum = song("zz", artist: "A", album: "X")
        let items = LibraryGrouping.albums(from: [noNum, d2, t3, t1])[0].items
        XCTAssertEqual(items.map { $0.metadata?.title }, ["b", "c", "zz", "a"].map(Optional.init))
    }

    func testAlbumsSortIgnoringLeadingArticleThenYear() {
        let the = song("1", artist: "A", album: "The Bends", year: "1995")
        let abc = song("2", artist: "A", album: "Amnesiac", year: "2001")
        let z = song("3", artist: "A", album: "Zeta")
        XCTAssertEqual(LibraryGrouping.albums(from: [z, the, abc]).map(\.title), ["Amnesiac", "The Bends", "Zeta"])
    }

    func testAlbumCoverIsFirstAvailable() {
        let cover = Data([0xFF, 0xD8])
        let noCover = song("1", artist: "A", album: "X", track: 1)
        let withCover = song("2", artist: "A", album: "X", track: 2, cover: cover)
        XCTAssertEqual(LibraryGrouping.albums(from: [noCover, withCover])[0].coverArtData, cover)
    }

    func testArtistSummaryCounts() {
        let a = song("1", artist: "G", album: "A1"), b = song("2", artist: "G", album: "A2"), c = song("3", artist: "G")
        let artist = LibraryGrouping.artists(from: [a, b, c])[0]
        XCTAssertEqual(artist.summary, "2 álbumes, 3 canciones")
        XCTAssertEqual(artist.albums.map(\.title), ["A1", "A2", LibraryGrouping.unknownAlbumTitle])
    }

    func testSortNameStripsArticles() {
        XCTAssertEqual(LibraryGrouping.sortName("The Beatles"), "Beatles")
        XCTAssertEqual(LibraryGrouping.sortName("Los Fabulosos Cadillacs"), "Fabulosos Cadillacs")
        XCTAssertEqual(LibraryGrouping.sortName("Alaska"), "Alaska", "no confunde 'A' con artículo dentro de la palabra")
        XCTAssertEqual(LibraryGrouping.sortName("The"), "The")
        XCTAssertEqual(LibraryGrouping.sortName("…Little Broken Hearts"), "Little Broken Hearts")
        XCTAssertEqual(LibraryGrouping.sortName("'Plastic Beach' Instrumentals"), "Plastic Beach' Instrumentals")
        XCTAssertEqual(LibraryGrouping.sortName("(What's the Story) Morning Glory?"), "What's the Story) Morning Glory?")
        XCTAssertEqual(LibraryGrouping.sortName("..."), "...")
    }

    // MARK: - videoCollections (PLAN-biblioteca-medios-v2.md §3.4, Tanda 4)

    private func video(_ title: String, category: String, seriesName: String? = nil,
                        season: Int? = nil, episode: Int? = nil, year: String? = nil,
                        cover: Data? = nil) -> AuraStudio.LibraryItem {
        var item = AuraStudio.LibraryItem(sourceURL: URL(fileURLWithPath: "/tmp/\(UUID().uuidString).mkv"))
        item.metadata = TrackMetadata(title: title, year: year, coverArtData: cover)
        item.category = category
        item.seriesName = seriesName
        item.season = season
        item.episode = episode
        return item
    }

    func testVideoCollectionsIgnoresNonMovieNonSeriesCategories() {
        let clip = video("Clip", category: "Videos")
        let photo = AuraStudio.LibraryItem(sourceURL: URL(fileURLWithPath: "/tmp/x.jpg"))
        XCTAssertTrue(LibraryGrouping.videoCollections(from: [clip, photo]).isEmpty)
    }

    func testStandaloneMovieBecomesItsOwnGroup() {
        let movie = video("Little Amelie", category: "Películas", year: "2025")
        let groups = LibraryGrouping.videoCollections(from: [movie])
        XCTAssertEqual(groups.count, 1)
        XCTAssertEqual(groups[0].title, "Little Amelie")
        XCTAssertEqual(groups[0].year, "2025")
        XCTAssertFalse(groups[0].isSeries)
        XCTAssertTrue(groups[0].seasons.isEmpty)
        XCTAssertEqual(groups[0].episodeCount, 1)
    }

    func testSeriesWithTwoSeasonsGroupsByNormalizedSeriesName() {
        let e1 = video("Piloto", category: "Series", seriesName: "Mi Serie", season: 1, episode: 1)
        let e2 = video("Segundo", category: "Series", seriesName: "mi serie", season: 1, episode: 2)
        let e3 = video("Estreno T2", category: "Series", seriesName: "Mi Serie", season: 2, episode: 1)
        let groups = LibraryGrouping.videoCollections(from: [e1, e2, e3])
        XCTAssertEqual(groups.count, 1, "misma serie, aunque la grafía del tag difiera en mayúsculas")
        let show = groups[0]
        XCTAssertTrue(show.isSeries)
        XCTAssertEqual(show.episodeCount, 3)
        XCTAssertEqual(show.seasons.map(\.number), [1, 2])
        XCTAssertEqual(show.seasons[0].items.map { $0.episode }, [1, 2], "episodios ordenados dentro de la temporada")
    }

    func testEpisodeWithoutSeasonNumberGoesLast() {
        let withSeason = video("Ep 1", category: "Series", seriesName: "Serie X", season: 1, episode: 1)
        let withoutSeason = video("Extra", category: "Series", seriesName: "Serie X", season: nil, episode: nil)
        let show = LibraryGrouping.videoCollections(from: [withSeason, withoutSeason])[0]
        XCTAssertEqual(show.seasons.map(\.number), [1, VideoCollectionGroup.noSeasonNumber],
                       "el cajón 'Sin temporada' siempre va al final, sin importar el orden de entrada")
    }

    func testMovieTitlesIgnoreLeadingArticleWhenSorting() {
        let a = video("The Matrix", category: "Películas")
        let b = video("Amelie", category: "Películas")
        let groups = LibraryGrouping.videoCollections(from: [a, b])
        XCTAssertEqual(groups.map(\.title), ["Amelie", "The Matrix"], "'The' no cuenta para el orden alfabético")
    }

    func testSeriesEnglishCategoryDisplayNameAlsoGroups() {
        // D-283: item.category se guarda como el displayName LOCALIZADO
        // -- "Series" en inglés (coincide con el español acá, pero
        // MediaCategory.series.displayNameEnglish es la fuente real).
        let e = video("Ep", category: MediaCategory.series.displayNameEnglish, seriesName: "Show", season: 1, episode: 1)
        let groups = LibraryGrouping.videoCollections(from: [e])
        XCTAssertEqual(groups.count, 1)
        XCTAssertTrue(groups[0].isSeries)
    }

    func testMoviePosterComesFromCoverArtData() {
        let data = Data("poster".utf8)
        let movie = video("X", category: "Películas", cover: data)
        XCTAssertEqual(LibraryGrouping.videoCollections(from: [movie])[0].posterData, data)
    }
}

File: studio/windows/tests/AuraStudio.Core.Tests/LibraryGroupingTests.cs
using AuraStudio.Core;
using AuraStudio.Core.Library;
using Xunit;

namespace AuraStudio.Core.Tests;

/// <summary>
/// Las cuadrículas de Álbumes, Artistas, Películas, Series y álbumes de fotos
/// (ST-031). Son grupos <b>en memoria</b>: nada de esto crea carpetas ni cambia
/// la organización en disco.
/// </summary>
public class LibraryGroupingTests
{
    private static LibraryItem Song(
        string path, string? title = null, string? artist = null, string? albumArtist = null,
        string? album = null, string? genre = null, string? year = null,
        int? track = null, int? disc = null, byte[]? cover = null, bool favorite = false,
        double? duration = null)
        => new()
        {
            SourcePath = path,
            Kind = LibraryItemKind.Music,
            Metadata = new TrackMetadata
            {
                Title = title,
                Artist = artist,
                AlbumArtist = albumArtist,
                Album = album,
                Genre = genre,
                Year = year,
                TrackNumber = track,
                DiscNumber = disc,
                CoverArtData = cover,
                IsFavorite = favorite,
                DurationSeconds = duration
            }
        };

    private static LibraryItem Video(
        string path, string category, string? title = null, string? series = null,
        int? season = null, int? episode = null, string? year = null, byte[]? poster = null)
        => new()
        {
            SourcePath = path,
            Kind = LibraryItemKind.Video,
            Category = category,
            SeriesName = series,
            Season = season,
            Episode = episode,
            Metadata = new TrackMetadata { Title = title, Year = year, CoverArtData = poster }
        };

    private static LibraryItem Photo(string path, string category, string? album = null) =>
        new() { SourcePath = path, Kind = LibraryItemKind.Photo, Category = category, PhotoAlbum = album };

    // MARK: - Normalización

    [Fact]
    public void TwoWaysOfWritingTheSameAlbumLandInOneGroup()
    {
        // "Álbum" y "album " son el mismo álbum; si no, la cuadrícula mostraría
        // el mismo disco dos veces.
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Signos", artist: "Soda Stereo"),
            Song(@"C:\m\2.mp3", album: "signos ", artist: "soda stereo")
        ]);

        AlbumGroup album = Assert.Single(albums);
        Assert.Equal(2, album.TrackCount);
    }

    [Fact]
    public void TheSpellingShownIsTheOneOfTheFirstTrackThatCameIn()
    {
        // Así una grafía descuidada que llegue después no le cambia el nombre al
        // álbum entero.
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Signos", artist: "Soda Stereo"),
            Song(@"C:\m\2.mp3", album: "SIGNOS", artist: "Soda Stereo")
        ]);

        Assert.Equal("Signos", albums[0].Title);
    }

    [Fact]
    public void AnAlbumIsGroupedByItsAlbumArtistNotByEachTracksArtist()
    {
        // Misma precedencia que la ruta de sincronización: lo que se ve acá
        // tiene que coincidir con las carpetas del iPod.
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Varios", artist: "Invitado 1", albumArtist: "La Banda"),
            Song(@"C:\m\2.mp3", album: "Varios", artist: "Invitado 2", albumArtist: "La Banda")
        ]);

        AlbumGroup album = Assert.Single(albums);
        Assert.Equal("La Banda", album.Artist);
    }

    // MARK: - Álbumes

    [Fact]
    public void AlbumsWithoutATitleGoIntoSinAlbumAtTheEnd()
    {
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", artist: "Soda Stereo"),
            Song(@"C:\m\2.mp3", album: "Signos", artist: "Soda Stereo")
        ]);

        Assert.Equal("Signos", albums[0].Title);
        Assert.Equal(LibraryGrouping.UnknownAlbumTitle, albums[^1].Title);
        Assert.True(albums[^1].IsUnknown);
    }

    [Fact]
    public void TheInitialArticleIsIgnoredWhenOrdering()
    {
        // Como Music.app: "Los Fabulosos Cadillacs" va en la F.
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Zamba", artist: "X"),
            Song(@"C:\m\2.mp3", album: "El Amor", artist: "X"),
            Song(@"C:\m\3.mp3", album: "Bailando", artist: "X")
        ]);

        Assert.Equal(["El Amor", "Bailando", "Zamba"], albums.Select(a => a.Title));
    }

    [Theory]
    [InlineData("The Wall", "Wall")]
    [InlineData("Los Fabulosos", "Fabulosos")]
    [InlineData("…Little Broken Hearts", "Little Broken Hearts")]
    [InlineData("(What's the Story)", "What's the Story)")]
    [InlineData("El", "El")]                // el artículo solo NO se vacía
    [InlineData("...", "...")]              // puro signo tampoco
    public void TheSortNameDropsArticlesAndLeadingPunctuation(string name, string expected)
        => Assert.Equal(expected, LibraryGrouping.SortName(name));

    [Fact]
    public void TheAlbumCoverIsTheFirstTrackThatHasOne()
    {
        byte[] cover = [1, 2, 3];
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Signos", artist: "X", track: 1),
            Song(@"C:\m\2.mp3", album: "Signos", artist: "X", track: 2, cover: cover)
        ]);

        // ST-208: el grupo apunta a la PISTA que tiene la tapa, no a sus bytes.
        Assert.Equal(cover, albums[0].CoverItem?.Metadata?.CoverArtData);
    }

    [Fact]
    public void TheCardSaysHowManySongsAndFromWhatYear()
    {
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Signos", artist: "X", year: "1986"),
            Song(@"C:\m\2.mp3", album: "Signos", artist: "X", year: "1986")
        ]);

        Assert.Equal("2 canciones · 1986", albums[0].SubtitleDetail);
    }

    [Fact]
    public void ASingleSongIsWrittenInTheSingular()
    {
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
            [Song(@"C:\m\1.mp3", album: "Sencillo", artist: "X")]);

        Assert.Equal("1 canción", albums[0].SubtitleDetail);
    }

    [Fact]
    public void AnAlbumIsFavoriteIfAnyOfItsSongsIs()
    {
        IReadOnlyList<AlbumGroup> albums = LibraryGrouping.Albums(
        [
            Song(@"C:\m\1.mp3", album: "Signos", artist: "X"),
            Song(@"C:\m\2.mp3", album: "Signos", artist: "X", favorite: true)
        ]);

        Assert.True(albums[0].IsFavorite);
    }

    // MARK: - Orden de las pistas

    [Fact]
    public void TracksAreOrderedByDiscThenTrackThenTitle()
    {
        IReadOnlyList<LibraryItem> tracks = LibraryGrouping.SortedTracks(
        [
            Song(@"C:\m\d2t1.mp3", title: "D2T1", disc: 2, track: 1),
            Song(@"C:\m\d1t2.mp3", title: "D1T2", disc: 1, track: 2),
            Song(@"C:\m\d1t1.mp3", title: "D1T1", disc: 1, track: 1)
        ]);

        Assert.Equal(["D1T1", "D1T2", "D2T1"], tracks.Select(t => t.DisplayTitle));
    }

    [Fact]
    public void ATrackWithoutADiscNumberCountsAsDiscOne()
    {
        // Como Music.app: sin ese dato no puede quedar antes de TODO el disco 1.
        IReadOnlyList<LibraryItem> tracks = LibraryGrouping.SortedTracks(
        [
            Song(@"C:\m\a.mp3", title: "sin disco", track: 5),
            Song(@"C:\m\b.mp3", title: "disco 1", disc: 1, track: 1),
            Song(@"C:\m\c.mp3", title: "disco 2", disc: 2, track: 1)
        ]);

        Assert.Equal(["disco 1", "sin disco", "disco 2"], tracks.Select(t => t.DisplayTitle));
    }

    [Fact]
    public void ATrackWithoutATrackNumberGoesToTheEndOfItsDisc()
    {
        // Ahí no hay un valor razonable que suponer, a diferencia del disco.
        IReadOnlyList<LibraryItem> tracks = LibraryGrouping.SortedTracks(
        [
            Song(@"C:\m\a.mp3", title: "sin número"),
            Song(@"C:\m\b.mp3", title: "pista 9", track: 9)
        ]);

        Assert.Equal(["pista 9", "sin número"], tracks.Select(t => t.DisplayTitle));
    }

    // MARK: - Artistas

    [Fact]
    public void AnArtistCollectsItsAlbumsAndCountsItsSongs()
    {
        IReadOnlyList<ArtistGroup> artists = LibraryGrouping.Artists(
        [
            Song(@"C:\m\1.mp3", album: "Signos", artist: "Soda Stereo"),
            Song(@"C:\m\2.mp3", album: "Signos", artist: "Soda Stereo"),
            Song(@"C:\m\3.mp3", album: "Nada Personal", artist: "Soda Stereo")
        ]);

        ArtistGroup artist = Assert.Single(artists);
        Assert.Equal(2, artist.Albums.Count);
        Assert.Equal(3, artist.TrackCount);
        Assert.Equal("2 álbumes, 3 canciones", artist.Summary);
    }

    [Fact]
    public void AnArtistWithOnlyLooseSongsIsSummarizedWithoutAlbums()
    {
        IReadOnlyList<ArtistGroup> artists = LibraryGrouping.Artists(
            [Song(@"C:\m\1.mp3", artist: "Soda Stereo")]);

        Assert.Equal("1 canción", artists[0].Summary);
    }

    [Fact]
    public void SongsWithoutAnArtistEndUpInArtistaDesconocidoAtTheEnd()
    {
        IReadOnlyList<ArtistGroup> artists = LibraryGrouping.Artists(
        [
            Song(@"C:\m\1.mp3", album: "Suelto"),
            Song(@"C:\m\2.mp3", album: "Signos", artist: "Soda Stereo")
        ]);

        Assert.Equal("Soda Stereo", artists[0].Name);
        Assert.Equal(LibraryGrouping.UnknownArtistName, artists[^1].Name);
        Assert.True(artists[^1].IsUnknown);
    }

    [Fact]
    public void AnArtistFallsBackToTheCoverOfItsFirstAlbumThatHasOne()
    {
        byte[] cover = [9, 9];
        IReadOnlyList<ArtistGroup> artists = LibraryGrouping.Artists(
        [
            Song(@"C:\m\1.mp3", album: "A", artist: "X"),
            Song(@"C:\m\2.mp3", album: "B", artist: "X", cover: cover)
        ]);

        Assert.Equal(cover, artists[0].FallbackCoverItem?.Metadata?.CoverArtData);
    }

    // MARK: - Películas y series

    [Fact]
    public void EveryEpisodeOfASeriesIsOneGroupNotOnePerFile()
    {
        IReadOnlyList<VideoCollectionGroup> collections = LibraryGrouping.VideoCollections(
        [
            Video(@"C:\v\1.mkv", "Series", series: "Chespirito", season: 1, episode: 1),
            Video(@"C:\v\2.mkv", "Series", series: "Chespirito", season: 1, episode: 2),
            Video(@"C:\v\3.mkv", "Series", series: "Chespirito", season: 2, episode: 1)
        ]);

        VideoCollectionGroup series = Assert.Single(collections);
        Assert.True(series.IsSeries);
        Assert.Equal("Chespirito", series.Title);
        Assert.Equal(3, series.EpisodeCount);
        Assert.Equal(2, series.Seasons.Count);
    }

    [Fact]
    public void EpisodesWithoutASeasonGoIntoTheirOwnDrawerAtTheEnd()
    {
        IReadOnlyList<VideoCollectionGroup> collections = LibraryGrouping.VideoCollections(
        [
            Video(@"C:\v\x.mkv", "Series", series: "Chespirito", episode: 1),
            Video(@"C:\v\1.mkv", "Series", series: "Chespirito", season: 1, episode: 1)
        ]);

        IReadOnlyList<SeasonGroup> seasons = collections[0].Seasons;
        Assert.Equal(1, seasons[0].Number);
        Assert.Equal(VideoCollectionGroup.NoSeasonNumber, seasons[^1].Number);
    }

    [Fact]
    public void EpisodesInsideASeasonAreOrderedByNumber()
    {
        IReadOnlyList<VideoCollectionGroup> collections = LibraryGrouping.VideoCollections(
        [
            Video(@"C:\v\3.mkv", "Series", title: "tres", series: "S", season: 1, episode: 3),
            Video(@"C:\v\1.mkv", "Series", title: "uno", series: "S", season: 1, episode: 1)
        ]);

        Assert.Equal(["uno", "tres"], collections[0].Seasons[0].Items.Select(i => i.DisplayTitle));
    }

    [Fact]
    public void AMovieDoesNotGetSeasons()
    {
        IReadOnlyList<VideoCollectionGroup> collections = LibraryGrouping.VideoCollections(
            [Video(@"C:\v\peli.mp4", "Películas", title: "La Peli", year: "1999")]);

        Assert.False(collections[0].IsSeries);
        Assert.Empty(collections[0].Seasons);
        Assert.Equal("La Peli", collections[0].Title);
    }

    [Fact]
    public void TwoImportsOfTheSameMovieAreOneGroup()
    {
        IReadOnlyList<VideoCollectionGroup> collections = LibraryGrouping.VideoCollections(
        [
            Video(@"C:\v\a.mp4", "Películas", title: "La Peli"),
            Video(@"C:\v\b.mkv", "Películas", title: "la peli ")
        ]);

        Assert.Equal(2, Assert.Single(collections).EpisodeCount);
    }

    [Fact]
    public void TwoMoviesWithoutATitleNeverGetMergedTogether()
    {
        // Sin título se agrupan por su propio id: mezclarlas sería peor que
        // mostrarlas separadas.
        IReadOnlyList<VideoCollectionGroup> collections = LibraryGrouping.VideoCollections(
        [
            Video(@"C:\v\a.mp4", "Películas"),
            Video(@"C:\v\b.mp4", "Películas")
        ]);

        Assert.Equal(2, collections.Count);
    }

    [Fact]
    public void ACatalogWrittenInEnglishIsStillRecognized()
    {
        // D-228: la categoría se guarda como nombre visible, y la app de macOS
        // en inglés escribe "Movies". Tratarlo como desconocido dejaría esas
        // películas fuera de la vista.
        Assert.Single(LibraryGrouping.VideoCollections(
            [Video(@"C:\v\a.mp4", "Movies", title: "La Peli")]));

        Assert.True(MediaCategoryNames.IsSeriesCategory("Series"));
        Assert.False(MediaCategoryNames.IsSeriesCategory("Videos"));
    }

    [Fact]
    public void APlainVideoIsNotAMovieNorASeries()
    {
        // La vista de Películas/Series no muestra los videos sueltos.
        Assert.Empty(LibraryGrouping.VideoCollections(
            [Video(@"C:\v\clip.mp4", "Videos", title: "Un clip")]));
    }

    // MARK: - Álbumes de fotos

    [Fact]
    public void PhotoAlbumsLiveInsideOneCollectionAtATime()
    {
        // La categoría entra en la clave: "Fotos" e "Imágenes" pueden tener cada
        // una un álbum llamado igual sin que se mezclen.
        LibraryItem[] items =
        [
            Photo(@"C:\f\1.jpg", "Fotos", "Viaje"),
            Photo(@"C:\f\2.jpg", "Imágenes", "Viaje")
        ];

        Assert.Single(Assert.Single(LibraryGrouping.PhotoAlbums(items, "Fotos")).Items);
        Assert.Single(Assert.Single(LibraryGrouping.PhotoAlbums(items, "Imágenes")).Items);
    }

    [Fact]
    public void PhotosWithoutAnAlbumEndUpInSinAlbumAtTheEnd()
    {
        IReadOnlyList<PhotoAlbumGroup> albums = LibraryGrouping.PhotoAlbums(
        [
            Photo(@"C:\f\1.jpg", "Fotos"),
            Photo(@"C:\f\2.jpg", "Fotos", "Viaje")
        ], "Fotos");

        Assert.Equal("Viaje", albums[0].Title);
        Assert.Equal(LibraryGrouping.UnknownPhotoAlbumTitle, albums[^1].Title);
        Assert.True(albums[^1].IsUnknown);
    }

    [Fact]
    public void TheCardShowsUpToFourPhotosForItsMosaic()
    {
        IReadOnlyList<PhotoAlbumGroup> albums = LibraryGrouping.PhotoAlbums(
            [.. Enumerable.Range(1, 6).Select(i => Photo($@"C:\f\{i}.jpg", "Fotos", "Viaje"))],
            "Fotos");

        Assert.Equal(6, albums[0].Count);
        Assert.Equal(4, albums[0].PreviewPaths.Count);
    }

    [Fact]
    public void TheMosaicPrefersThePreparedFileOverTheOriginal()
    {
        var photo = Photo(@"C:\f\1.heic", "Fotos", "Viaje");
        photo.PreparedPath = @"C:\biblioteca\.preparados\1.jpg";

        IReadOnlyList<PhotoAlbumGroup> albums = LibraryGrouping.PhotoAlbums([photo], "Fotos");

        Assert.Equal(@"C:\biblioteca\.preparados\1.jpg", albums[0].PreviewPaths[0]);
    }

    // MARK: - Alcance

    [Fact]
    public void AnEmptyLibraryGroupsIntoNothing()
    {
        Assert.Empty(LibraryGrouping.Albums([]));
        Assert.Empty(LibraryGrouping.Artists([]));
        Assert.Empty(LibraryGrouping.VideoCollections([]));
        Assert.Empty(LibraryGrouping.PhotoAlbums([], "Fotos"));
    }

    [Fact]
    public void AlbumsOnlyLookAtMusic()
    {
        Assert.Empty(LibraryGrouping.Albums(
            [Photo(@"C:\f\1.jpg", "Fotos"), Video(@"C:\v\1.mp4", "Películas", title: "x")]));
    }
}
